File: leetcode/algo2/day3/15_3sum.js
var threeSum = function (nums) {
  var res = [];
  if (nums.length < 3) return res;
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length - 2; i++) {
    if (i > 0 && nums[i] == nums[i - 1]) continue;
    let j = i + 1;
    let k = nums.length - 1;
    let target = 0 - nums[i];

    while (j < k) {
      let sum = nums[i] + nums[j];
      if (sum == target) {
        let temp = [nums[i], nums[j], nums[k]];
        res.push(temp);
        while (j < k && nums[j] == temp[1]) j++;
        while (j < k && nums[k] == temp[2]) k--;
      } else if (sum < target) j++;
      else k--;
    }
  }
  return res;
};

var nums = [-1, 0, 1, 2, -1, -4];
console.log(threeSum(nums));

File: leetcode/algo2/day3/15_3sum.cpp
#include <vector>
#include <iostream>
#include <algorithm>
std::vector<std::vector<int>> threeSum(std::vector<int> &nums)
{
    std::vector<std::vector<int>> result;
    if (nums.size() < 3)
        return result;
    std::sort(nums.begin(), nums.end()); // -4, -1, -1, 0, 1, 2
    for (int i = 0; i < nums.size() - 2; i++)
    {
        if (i > 0 && (nums[i] == nums[i - 1]))
            continue;
        int j = i + 1;
        int k = nums.size() - 1;
        int target = 0 - nums[i];
        while (j < k)
        {
            int sum_jk = nums[j] + nums[k];
            if (sum_jk == target)
            {
                result.push_back({nums[i], nums[j], nums[k]});
                std::cout << i << " " << j << " " << k << std::endl;
                j++;
                k--;
            }
            else if (sum_jk < target)
                j++;
            else
                k--;
        }
    }
    return result;
}
int main()
{
    std::vector<int> nums = {0, 0, 0, 0};
    for (auto x : threeSum(nums))
    {
        for (auto y : x)
            std::cout << y << " ";
        std::cout << std::endl;
    }
}